### PR TITLE
Generate OpenAPI definition

### LINKED ISSUE #1686: Create method to generate an openapi 3 operation for each method
OpenAPI (aka Swagger) defines a format to formally document REST APIs. All the information to generate these docs are actually available in protocol. 

In this ticket a generate function should generate openapi compatible object structure for all *typedmehods*. This object structure should consist of pydantic objects that can be cleanly serialized to json or yaml. 

The fastapi library already does this using pydantic. As a reference:

https://github.com/tiangolo/fastapi/blob/master/fastapi/openapi/
https://github.com/tiangolo/fastapi/blob/master/fastapi/openapi/models.py

However, we do not want to add a dependency to fastapi, so we have to add our own model. For now, only add the attributes required for the methods.

As for the types of parameters and the return value: either use the type of the annotation or when it is a subclass of BaseModel, use object (dict) for now. Another ticket adds jsonschema based on pydantic schema support.

### PR DESCRIPTION
**Pull request by andraskvr**
Closes #1686

### CODE REVIEW

File: src/inmanta/protocol/openapi/converter.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    def get_path_of_method(self, methods: Dict[str, UrlMethod]) -> str:
+        path = [
+            f"/{method.properties._api_prefix}/v{method.properties.api_version}{method.properties._path.path}".replace(
+                "<", "{"
+            ).replace(">", "}")
+            for method in methods.values()
+        ][0]
+        return path
+
+    def extract_parameters_from_arg_options(self, response: Response, path: str, arg_options: Dict[str, ArgOption]):
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: 1. return type is missing
2. is it correct that some results are returned via the return value and some via the response argument? This is confusing, it should either be avoided or documented

File: src/inmanta/protocol/openapi/converter.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        parameters = []
+        for option_name, option in arg_options.items():
+            param_dict = {}
+            if option.header:
+                param_dict["in"] = ParameterType.header
+                param_dict["name"] = option.header
+                param_dict["schema"] = {"type": "string"}
+                if option.reply_header:
+                    response.headers = {option.header: {"description": option.header, "schema": {"type": "string"}}}
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: no name field here? Why is the format so different compared to the param_dict?

File: src/inmanta/protocol/openapi/converter.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+from inmanta.server import config
+from inmanta.util import get_compiler_version
+
+
+def openapi_json_encoder(o):
+    if isinstance(o, BaseModel):
+        return o.dict(by_alias=True, exclude_none=True)
+
+
+class OpenApiConverter:
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: What is the main entry point of this class?

File: src/inmanta/protocol/openapi/converter.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                parameters.append(
+                    Parameter(
+                        **{"name": parameter_name, "in": ParameterType.path, "required": True, "schema": type_description}
+                    )
+                )
+        return properties
+
+    def get_openapi_type(self, parameter_type: Type):
+        # TODO handle inmanta types
+        if hasattr(parameter_type, "__name__"):
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: this logic scares me a bit, it should be documented

File: src/inmanta/protocol/openapi/converter.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        else:
+            type_name = "dict"
+        python_to_openapi_types_without_format = {
+            "bool": "boolean",
+            "int": "integer",
+            "str": "string",
+            "dict": "object",
+            "list": "array",
+        }
+        openapi_type_name = python_to_openapi_types_without_format.get(type_name)
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: 1. why match on name, not on type or subtype
2. Wouldn't it be better to explicitly get pydantic types(which all have a common ancestor class) out (either to handle them or ignore them)

File: src/inmanta/protocol/openapi/converter.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                return_properties = {url_method.properties.envelope_key: {"type": "object"}}
+            content = {"application/json": {"schema": {"type": "object", "properties": return_properties}}}
+            response.content = content
+
+    def get_request_body(self, properties: Dict):
+        request_body_dict = {
+            "required": True,
+            "content": {"application/json": {"schema": {"type": "object", "properties": properties}}},
+        }
+        request_body = RequestBody(**request_body_dict)
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: why kwargs when the dict is constructed above?

File: src/inmanta/protocol/openapi/model.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+from enum import Enum
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/model.py]: add an attribution

Comment (wouterdb) [reply src/inmanta/protocol/openapi/model.py]: and refer to the spec

File: src/inmanta/protocol/openapi/converter.py (hunk 2 above)
Comment (andraskvr) [reply src/inmanta/protocol/openapi/converter.py]: The name is the key in this dict:
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#header-object
So a Header is almost the same as a Parameter, but not exactly

File: src/inmanta/protocol/openapi/converter.py [hunk 8]
```diff
@@ -15,45 +15,72 @@
...
+        self.arg_option_handler = ArgOptionHandler()
+        self.type_converter = OpenApiTypeConverter()
+
+    def _get_servers(self) -> List[Dict[str, str]]:
         bind_port = config.get_bind_port()
         bind_addresses = config.server_bind_address.get()
         return [{"url": f"http://{bind_address}:{bind_port}/"} for bind_address in bind_addresses]
 
-    def get_top_level_object(self):
+    def _get_envelope(self) -> Dict[str, Dict]:
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: Why is this a dict and not an OpenAPI object?

File: src/inmanta/protocol/openapi/converter.py [hunk 9]
```diff
@@ -15,45 +15,72 @@
...
-        servers = self.get_servers()
-        return {"openapi": "3.0.2", "info": info, "paths": {}, "servers": servers}
+        servers = self._get_servers()
+        paths = {}
+        for methods in self.global_url_map.values():
+            api_methods = self._filter_api_methods(methods)
+            if len(api_methods) > 0:
+                path = self._get_path_of_method(api_methods)
+                path_item = self._extract_operations_from_methods(api_methods, path)
+                paths[path] = path_item
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: shouldn't None values be filtered out?

File: src/inmanta/protocol/openapi/converter.py [hunk 10]
```diff
@@ -15,45 +15,72 @@
...
+    def _filter_api_methods(self, methods: Dict[str, UrlMethod]) -> Dict[str, UrlMethod]:
         return {
             method_name: url_method
             for method_name, url_method in methods.items()
             if "api" in url_method.properties.client_types
         }
 
-    def get_path_of_method(self, methods: Dict[str, UrlMethod]) -> str:
+    def _get_path_of_method(self, methods: Dict[str, UrlMethod]) -> str:
         path = [
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: isn't this duplicate code wrt protocol? I expect it to be somewhere int there as well?

File: src/inmanta/protocol/openapi/converter.py [hunk 11]
```diff
@@ -62,124 +89,198 @@ def get_path_of_method(self, methods: Dict[str, UrlMethod]) -> str:
...
+    def convert_function_params_to_query_params(self) -> List[Parameter]:
+        parameters: List[Parameter] = []
+        for parameter_name, parameter_type in self.non_path_params.items():
+            type_description = self.type_converter.get_openapi_type(parameter_type)
+            parameters.append(Parameter(**{"name": parameter_name, "in": ParameterType.query, "schema": type_description}))
+        return parameters
+
+    def _convert_function_params_to_openapi_request_body_properties(
+        self, function_parameters: Dict[str, inspect.Parameter]
+    ) -> Dict[str, Dict[str, str]]:
```
Comment (wouterdb) [inline src/inmanta/protocol/openapi/converter.py]: the inner dict, doesn't that correspond to a more exact model type?

File: src/inmanta/protocol/openapi/converter.py (hunk 9 above)
Comment (wouterdb) [reply src/inmanta/protocol/openapi/converter.py]: this still produces a type error

**Pull request merged**